### ISSUE TITLE
Hey Kevin - fyi - VASA-1 paper recreation

### ISSUE THREAD

**Issue by johndpope** [Kevinfringe/MegaPortrait#5]
the whitepaper is mentioned in microsoft's latest VASA paper
https://www.microsoft.com/en-us/research/project/vasa-1/

I used claude ai - opus / chatgpt to help me attempt to code out EMOTE - portrait alive
https://github.com/johndpope/emote-hack
(since abandoned)

I'm looking at attempting the VASA paper - and getting this codebase to work may help.
Checkout Claude.ai - you can upload 5 files
you can give it a html version of whitepaper - your code in a single file - diagrams - and ask it to align your code to paper.

UPDATE - found this
https://github.com/Meta-Portrait/MetaPortrait

UPDATE
https://github.com/johndpope/vasa-1-hack

UPDATE

when I supply Claude the code from this repo  + 2 screenshots of white paper - it has this criticism

https://github.com/johndpope/VASA-1-hack/blob/main/reference/kevinfringe_MegaPortrait.py

![Screenshot from 2024-04-28 19-01-09](https://github.com/Kevinfringe/MegaPortrait/assets/289994/6129e9d0-cb4f-462b-9a80-a150e6dfef14)

Thank you for sharing these code snippets and MegaPortrait paper excerpts. Based on the provided information, here are a few suggestions to better align the code with the paper:

The paper mentions using a pre-trained VGG19 for calculating perceptual losses (section 3.2). The code currently uses vgg_face for the face loss, but does not include VGG19 for the perceptual loss. Consider adding VGG19 and calculating the perceptual loss as described.
The adversarial loss in the code uses a hinge loss, while the paper describes using a standard feature-matching loss (section 3.2) to improve stability. Consider replacing the hinge loss with the feature matching loss mentioned.
The code has separate Warp_G, G3d and G2d modules, but the paper describes them as W_s→γ, W_→d and a combined G3D→2D network. Consider combining G3d and G2d into one module to match the architecture diagram.
The paper describes training in two stages - first a low resolution model, then a high-resolution one (section 3.2). The current code does not reflect this staged training approach.
A cosine cycle consistency loss is included in the code, but not explicitly mentioned in the paper excerpts provided. Double check if this matches the full MegaPortrait paper.

In summary, the key areas to focus on are:

Adding VGG19 for perceptual loss - (THIS IS INCORRECT BY CLAUDE) 

![Screenshot from 2024-04-28 19-19-34](https://github.com/Kevinfringe/MegaPortrait/assets/289994/728f8279-bbc5-4fa1-8b4d-7ef023ee69c2)
Combining G3d and G2d - not sure if this is ideal?
![Screenshot from 2024-04-28 19-19-26](https://github.com/Kevinfringe/MegaPortrait/assets/289994/ff6a95b5-ff62-4025-8aee-15dfef858b0c)

TBC - 
Using feature-matching GAN loss
Implementing two-stage low res and high res training
Verifying the cosine loss matches the paper

related
https://github.com/johndpope/VASA-1-hack/issues/8

UPDATE - not sure how to resolve this - 

![Screenshot from 2024-04-29 06-30-25](https://github.com/Kevinfringe/MegaPortrait/assets/289994/a4f07947-c189-4e7a-8e4b-d5b3793e7332)

**Comment by johndpope** [comment]
@Jie-zju  found this error with your warping code - Claude has suggested fix.


![Screenshot from 2024-05-10 05-37-50](https://github.com/Kevinfringe/MegaPortrait/assets/289994/66b69205-9ce7-4e4f-ad7d-2306c5be5cb4)

![Screenshot from 2024-05-10 05-38-00](https://github.com/Kevinfringe/MegaPortrait/assets/289994/8e166486-6e4f-4ee4-a07b-11d89178ad46)

```python

import torch
import torch.nn.functional as F

def warp_3d(x, theta):
    # Generate 3D grid
    grid = F.affine_grid(theta, x.size())
    
    # Sample the input tensor using the grid
    warped_x = F.grid_sample(x, grid)
    
    return warped_x

# Replace cv.warpPerspective with warp_3d function
warp_3d_vs = warp_3d(v_s, W_rt_s)
warp_3d_vs = warp_3d(warp_3d_vs, W_em_s)

vs_d = warp_3d(warp_3d_vs, W_rt_d)
vs_d = warp_3d(vs_d, W_em_d)
```

**Comment by Jie-zju** [comment]
> @Jie-zju found this error with your warping code - Claude has suggested fix.
> 
> ![Screenshot from 2024-05-10 05-37-50](https://private-user-images.githubusercontent.com/289994/329366508-66b69205-9ce7-4e4f-ad7d-2306c5be5cb4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU0MTkyOTIsIm5iZiI6MTcxNTQxODk5MiwicGF0aCI6Ii8yODk5OTQvMzI5MzY2NTA4LTY2YjY5MjA1LTljZTctNGU0Zi1hZDdkLTIzMDZjNWJlNWNiNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxMVQwOTE2MzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iY2UwNDkyZDdhNDFkYmJhNTQ5OWM1MTFjMDI4NWRhOGFmYjUxMTUwY2JlZGEwNTA3MzAzY2Y3ZmZiZDIxYjdjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.641nNRMUms9_xCqCURC1TBtj0OANnUt_dfjQAJALKJM)
> 
> ![Screenshot from 2024-05-10 05-38-00](https://private-user-images.githubusercontent.com/289994/329366406-8e166486-6e4f-4ee4-a07b-11d89178ad46.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU0MTkyOTIsIm5iZiI6MTcxNTQxODk5MiwicGF0aCI6Ii8yODk5OTQvMzI5MzY2NDA2LThlMTY2NDg2LTZlNGYtNGVlNC1hMDdiLTExZDg5MTc4YWQ0Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxMVQwOTE2MzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYTlkZTMwYWU4N2Q1ZmJhZWU1MjExYjk4MmRmNmE3YzkyOGNmNGIwOWI1MDA5YmQwZTg1YTcyMTM1NzJkNzdkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.SB-vsmuB585aMaStWi3oWb4NqZiV2sdGXFySHkq8GXE)
> 
> ```python
> import torch
> import torch.nn.functional as F
> 
> def warp_3d(x, theta):
>     # Generate 3D grid
>     grid = F.affine_grid(theta, x.size())
>     
>     # Sample the input tensor using the grid
>     warped_x = F.grid_sample(x, grid)
>     
>     return warped_x
> 
> # Replace cv.warpPerspective with warp_3d function
> warp_3d_vs = warp_3d(v_s, W_rt_s)
> warp_3d_vs = warp_3d(warp_3d_vs, W_em_s)
> 
> vs_d = warp_3d(warp_3d_vs, W_rt_d)
> vs_d = warp_3d(vs_d, W_em_d)
> ```

@johndpope Thanks for your wonderful reply! I am looking forward your implementation!

**Comment by johndpope** [comment]
i started getting Claude to update the training code and now almost completely rebuilt it
https://github.com/johndpope/MegaPortrait/blob/master/train.py

https://github.com/johndpope/MegaPortrait/blob/master/model.py


![Screenshot from 2024-05-11 20-05-16](https://github.com/Kevinfringe/MegaPortrait/assets/289994/f0fd6dc0-7e84-45f0-b467-d783fd8a50b4)

UPDATE
Because there's many third party code  (samsung rome) / that needs merging - I create new project.
https://github.com/johndpope/MegaPortrait-hack
